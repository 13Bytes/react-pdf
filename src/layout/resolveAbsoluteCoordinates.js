import * as R from 'ramda';

/**
 * Translates relative coordinates to absolute coordinates
 *
 * @param {Object} root node
 * @returns {Object} root node with absolute coordinates
 */
const resolveAbsoluteCoordinates = node => {
  const top = R.pathOr(0, ['box', 'top'], node);
  const left = R.pathOr(0, ['box', 'left'], node);

  return R.evolve({
    children: R.map(
      R.compose(
        resolveAbsoluteCoordinates,
        R.evolve({
          box: R.evolve({
            top: R.add(top),
            bottom: R.add(top),
            left: R.add(left),
            right: R.add(left),
          }),
        }),
      ),
    ),
  })(node);
};

export default resolveAbsoluteCoordinates;
