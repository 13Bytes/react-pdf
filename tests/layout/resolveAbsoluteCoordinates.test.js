import resolveAbsoluteCoordinates from '../../src/layout/resolveAbsoluteCoordinates';

describe('layout resolveAbsoluteCoordinates', () => {
  test('should leave page coordinates as they are', () => {
    const root = {
      type: 'ROOT',
      children: [
        {
          type: 'DOCUMENT',
          children: [
            {
              type: 'PAGE',
              box: {
                top: 0,
                left: 0,
                width: 100,
                height: 100,
              },
            },
            {
              type: 'PAGE',
              box: {
                top: 0,
                left: 0,
                width: 200,
                height: 200,
              },
            },
          ],
        },
      ],
    };
    const result = resolveAbsoluteCoordinates(root);

    expect(result).toMatchSnapshot();
  });

  test('should leave page children coordinates as they are', () => {
    const root = {
      type: 'ROOT',
      children: [
        {
          type: 'DOCUMENT',
          children: [
            {
              type: 'PAGE',
              box: {
                top: 0,
                left: 0,
                width: 100,
                height: 100,
              },
              children: [
                {
                  type: 'VIEW',
                  box: {
                    top: 20,
                    left: 20,
                    width: 50,
                    height: 50,
                  },
                },
                {
                  type: 'VIEW',
                  box: {
                    top: 70,
                    left: 20,
                    width: 50,
                    height: 20,
                  },
                },
              ],
            },
          ],
        },
      ],
    };
    const result = resolveAbsoluteCoordinates(root);

    expect(result).toMatchSnapshot();
  });

  test('should resolve nested children coordinates', () => {
    const root = {
      type: 'ROOT',
      children: [
        {
          type: 'DOCUMENT',
          children: [
            {
              type: 'PAGE',
              box: {
                top: 0,
                left: 0,
                width: 100,
                height: 100,
              },
              children: [
                {
                  type: 'VIEW',
                  box: {
                    top: 20,
                    left: 20,
                    width: 50,
                    height: 50,
                  },
                  children: [
                    {
                      type: 'VIEW',
                      box: {
                        top: 0,
                        left: 0,
                        width: 40,
                        height: 30,
                      },
                    },
                    {
                      type: 'VIEW',
                      box: {
                        top: 20,
                        left: 20,
                        width: 10,
                        height: 10,
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    };
    const result = resolveAbsoluteCoordinates(root);

    expect(result).toMatchSnapshot();
  });
});
