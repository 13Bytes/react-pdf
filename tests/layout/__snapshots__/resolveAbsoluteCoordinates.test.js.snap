// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout resolveAbsoluteCoordinates should leave page children coordinates as they are 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 100,
        "left": 0,
        "top": 0,
        "width": 100,
      },
      "children": Array [
        Object {
          "box": Object {
            "height": 50,
            "left": 20,
            "top": 20,
            "width": 50,
          },
          "type": "VIEW",
        },
        Object {
          "box": Object {
            "height": 20,
            "left": 20,
            "top": 70,
            "width": 50,
          },
          "type": "VIEW",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveAbsoluteCoordinates should leave page coordinates as they are 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 100,
        "left": 0,
        "top": 0,
        "width": 100,
      },
      "type": "PAGE",
    },
    Object {
      "box": Object {
        "height": 200,
        "left": 0,
        "top": 0,
        "width": 200,
      },
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;

exports[`layout resolveAbsoluteCoordinates should resolve nested children coordinates 1`] = `
Object {
  "children": Array [
    Object {
      "box": Object {
        "height": 100,
        "left": 0,
        "top": 0,
        "width": 100,
      },
      "children": Array [
        Object {
          "box": Object {
            "height": 50,
            "left": 20,
            "top": 20,
            "width": 50,
          },
          "children": Array [
            Object {
              "box": Object {
                "height": 30,
                "left": 20,
                "top": 20,
                "width": 40,
              },
              "type": "VIEW",
            },
            Object {
              "box": Object {
                "height": 10,
                "left": 40,
                "top": 40,
                "width": 10,
              },
              "type": "VIEW",
            },
          ],
          "type": "VIEW",
        },
      ],
      "type": "PAGE",
    },
  ],
  "type": "DOCUMENT",
}
`;
